stdlib/public/Concurrency/AsyncStreamBackpressure.swift

```diff
@@ -0,0 +1,1961 @@
+            // This is just the references dropping afterwards.
+            return .none
+
+        case .modify:
+            fatalError("AsyncStream internal inconsistency")
+        }
+    }
+
+    /// Actions returned by `sourceDeinitialized()`.
+    enum SourceDeinitializedAction {
+        /// Indicates that `onTermination` should be called.
+        case callOnTermination((() -> Void)?)
+        /// Indicates that  all producers should be failed and `onTermination` should be called.
+        case failProducersAndCallOnTermination(
+            [(Result<Void, Error>) -> Void],
+            (@Sendable () -> Void)?
+        )
+        /// Indicates that all producers should be failed.
+        case failProducers([(Result<Void, Error>) -> Void])
+    }
+
+    mutating func sourceDeinitialized() -> SourceDeinitializedAction? {
+        switch self._state {
+        case .initial(let initial):
+            // The source got deinited before anything was written
+            self._state = .finished(iteratorInitialized: initial.iteratorInitialized)
+            return .callOnTermination(initial.onTermination)
+
+        case .streaming(let streaming):
+            if streaming.buffer.isEmpty {
+                // We can transition to finished right away since the buffer is empty now
+                self._state = .finished(iteratorInitialized: streaming.iteratorInitialized)
+
+                return .failProducersAndCallOnTermination(
+                    Array(streaming.producerContinuations.map { $0.1 }),
+                    streaming.onTermination
+                )
+            } else {
+                // The continuation must be `nil` if the buffer has elements
+                precondition(streaming.consumerContinuation == nil)
+
+                self._state = .sourceFinished(.init(
+                    iteratorInitialized: streaming.iteratorInitialized,
+                    buffer: streaming.buffer,
+                    failure: nil,
+                    onTermination: streaming.onTermination
+                ))
+
+                return .failProducers(
+                    Array(streaming.producerContinuations.map { $0.1 })
+                )
+            }
+
+        case .sourceFinished, .finished:
+            // This is normal and we just have to tolerate it
+            return .none
+
+        case .modify:
+            fatalError("AsyncStream internal inconsistency")
+        }
+    }
+
+    /// Actions returned by `write()`.
+    enum WriteAction {
+        /// Indicates that the producer should be notified to produce more.
+        case returnProduceMore
+        /// Indicates that the producer should be suspended to stop producing.
+        case returnEnqueue(
+            callbackToken: UInt
+        )
+        /// Indicates that the consumer should be resumed and the producer should be notified to produce more.
+        case resumeConsumerAndReturnProduceMore(
+            continuation: CheckedContinuation<Element?, Error>,
```

Comment:
probably should move on to `UnsafeContinuation` for landing this.

Reply:
Don't we want to bake this with `CheckedContinuation` for at least a release and then move to `UnsafeContinuation`?